File: pkcs11wrapper-tests/app/src/androidTest/java/ru/rutoken/pkcs11wrapper/main/Pkcs11TokenTest.java
package ru.rutoken.pkcs11wrapper.main;

import android.util.Log;

import org.junit.ClassRule;
import org.junit.Test;
import org.junit.rules.RuleChain;
import org.junit.rules.TestRule;

import ru.rutoken.pkcs11wrapper.constant.standard.Pkcs11ReturnValue;
import ru.rutoken.pkcs11wrapper.rule.highlevel.ModuleRule;
import ru.rutoken.pkcs11wrapper.rule.highlevel.SlotRule;
import ru.rutoken.pkcs11wrapper.rule.highlevel.TokenRule;

import static org.hamcrest.Matchers.greaterThan;
import static org.hamcrest.Matchers.greaterThanOrEqualTo;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.fail;

public class Pkcs11TokenTest {
    private static final ModuleRule sModule = new ModuleRule();
    private static final SlotRule sSlot = new SlotRule(sModule);
    private static final TokenRule sToken = new TokenRule(sSlot);
    @ClassRule
    public static final TestRule sRuleChain = RuleChain.outerRule(sModule).around(sSlot).around(sToken);

    @Test
    public void getTokenInfo() {
        final var info = sToken.getValue().getTokenInfo();
        assertEquals("Aktiv Co.                       ", info.getManufacturerId());
    }

    @Test
    public void getMechanismList() {
        final var mechanisms = sToken.getValue().getMechanismList();
        assertThat("no supported mechanisms", mechanisms.size(), greaterThan(0));
    }

    @Test
    public void getMechanismInfo() {
        for (var mechanism : sToken.getValue().getMechanismList()) {
            Log.d(getClass().getName(), "mechanism: " + mechanism);

            final var info = sToken.getValue().getMechanismInfo(mechanism);
            assertThat("min key size less than zero", info.getMinKeySize(), greaterThanOrEqualTo(0L));
            assertThat("max key size is less than min key size", info.getMaxKeySize(),
                    greaterThanOrEqualTo(info.getMinKeySize()));
        }
    }

    @Test
    public void openSession() {
        final Pkcs11Session session;
        try (var ignore = session = sToken.getValue().openSession(true)) {
        }
        checkSessionClosed(session);
    }

    @Test
    public void closeAllSessions() {
        final Pkcs11Session session1;
        final Pkcs11Session session2;
        try {
            session1 = sToken.getValue().openSession(true);
            session2 = sToken.getValue().openSession(true);
        } finally {
            sToken.getValue().closeAllSessions();
        }
        checkSessionClosed(session1);
        checkSessionClosed(session2);
    }

    private void checkSessionClosed(Pkcs11Session session) {
        try {
            session.getSessionInfo();
            fail("calling getSessionInfo() on closed session should throw");
        } catch (Pkcs11Exception e) {
            if (Pkcs11ReturnValue.CKR_SESSION_CLOSED != e.getCode() &&
                    Pkcs11ReturnValue.CKR_SESSION_HANDLE_INVALID != e.getCode())
                throw e;
        }
    }
}

File: pkcs11wrapper-tests/app/src/androidTest/java/ru/rutoken/pkcs11wrapper/main/Pkcs11TokenTest.kt
/*
 * Copyright (c) 2018, Aktiv-Soft JSC.
 * See the LICENSE file at the top-level directory of this distribution.
 * All Rights Reserved.
 */

package ru.rutoken.pkcs11wrapper.main

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.collections.shouldBeOneOf
import io.kotest.matchers.ints.shouldBeGreaterThan
import io.kotest.matchers.longs.shouldBeGreaterThanOrEqual
import io.kotest.matchers.string.shouldContain
import org.junit.ClassRule
import org.junit.Test
import org.junit.rules.RuleChain
import org.junit.rules.TestRule
import ru.rutoken.pkcs11wrapper.constant.standard.Pkcs11ReturnValue.CKR_SESSION_CLOSED
import ru.rutoken.pkcs11wrapper.constant.standard.Pkcs11ReturnValue.CKR_SESSION_HANDLE_INVALID
import ru.rutoken.pkcs11wrapper.constant.standard.Pkcs11UserType
import ru.rutoken.pkcs11wrapper.rule.highlevel.ModuleRule
import ru.rutoken.pkcs11wrapper.rule.highlevel.SlotRule
import ru.rutoken.pkcs11wrapper.rule.highlevel.TokenRule

class Pkcs11TokenTest {
    @Test
    fun getTokenInfo() {
        val info = token.value.tokenInfo
        info.manufacturerId shouldContain "Aktiv Co."
    }

    @Test
    fun getMechanismList() {
        val mechanisms = token.value.mechanismList
        mechanisms.size shouldBeGreaterThan 0
    }

    @Test
    fun getMechanismInfo() {
        for (mechanism in token.value.mechanismList) {
            val info = token.value.getMechanismInfo(mechanism)
            info.minKeySize shouldBeGreaterThanOrEqual 0
            info.maxKeySize shouldBeGreaterThanOrEqual info.minKeySize
        }
    }

    @Test
    fun openSession() {
        val session: Pkcs11Session
        token.value.openSession(true).use { session = it }
        checkSessionClosed(session)
    }

    @Test
    fun initToken() {
        token.value.apply {
            initToken(DEFAULT_ADMIN_PIN, TOKEN_LABEL)
            openSession(true).use { session ->
                session.login(Pkcs11UserType.CKU_SO, DEFAULT_ADMIN_PIN).use {
                    session.initPin(DEFAULT_USER_PIN)
                }
            }
            tokenInfo.label shouldContain TOKEN_LABEL
        }
    }

    @Test
    fun closeAllSessions() {
        val session1: Pkcs11Session
        val session2: Pkcs11Session

        token.value.apply {
            try {
                session1 = openSession(true)
                session2 = openSession(true)
            } finally {
                closeAllSessions()
            }
        }

        checkSessionClosed(session1)
        checkSessionClosed(session2)
    }

    private fun checkSessionClosed(session: Pkcs11Session) {
        val exception = shouldThrow<Pkcs11Exception> {
            session.sessionInfo
        }
        exception.code shouldBeOneOf listOf(CKR_SESSION_CLOSED, CKR_SESSION_HANDLE_INVALID)
    }

    companion object {
        private val module = ModuleRule()
        private val slot = SlotRule(module)
        private val token = TokenRule(slot)

        @ClassRule
        @JvmField
        val ruleChain: TestRule = RuleChain.outerRule(module).around(slot).around(token)
    }
}